### DISCUSSION TITLE
Controlling valves

### DISCUSSION THREAD

**Discussion by antamy** [tagyoureit/nodejs-poolController#687] [Q&A]
I have a basic setup - an Intellitouch i5+3, with an Aquacal heat pump heater (wired as a gas heater) and a solar array (simply activated by a valve). I would like to try and boost the heating and run both the solar array and the heater at the same time. The panel only lets me choose one heat source or a preferred heat source.

On the panel, there is a Configure Valves menu which shows Valve A - Used by Solar and Valve B - Unused.  In njspc, I see 3 valves: 

![image](https://user-images.githubusercontent.com/6841038/218273888-39aa4612-d374-45eb-bb0f-d8d82b501f5d.png)

What I'd like to be able to do it to set my heating to Heater and open the solar valve.  I can do this manually using the switch on the valve, but this is not ideal.

My ideal would be to increase the pump speed as well when doing this - normally I run my pump at 1800rpm for pool+heater, and 2500rpm for solar.

Is there any way to do any of this?

Thanks

**Comment by rstrouse** [comment]
Lets be certain here.  Do you have an IntelliTouch i5+3 or i5+3S?

EDIT:  This is a standard function but I am seeing 1 too many valves for an i5+3 (not S) so if you could generate a replay I can see what the configuration is that is returned from your panel.  It does not appear to have the normal valve packets.  If your panel is configured as you mentioned then Solar should appear for valve A.  And when you select the Solar heater or Solar Pref into the body as the heater it should automatically move the valve.   IntelliTouch will not allow you to run both at the same time but it will switch between the heatpump and the solar array with preference for heating by solar.

**Reply by antamy** [reply]
i5+3.  This is what njspc reports:

![image](https://user-images.githubusercontent.com/6841038/218274514-33845e0d-2e3a-4f1c-bd10-1f4482df0d39.png)

**Reply by rstrouse** [reply]
Can I get you to generate a replay so I can see the configuration?  Under logging press Capture Replay and check the box for Capture Configuration Reload.  When the status returns to ready press the End Capture button.

**Reply by antamy** [reply]
Here's the capture.  Not sure what "Invalid" is on this page:

![image](https://user-images.githubusercontent.com/6841038/218275436-4e14dfea-9db5-479f-9f64-e30315a4d367.png)

[replay(2).zip](https://github.com/tagyoureit/nodejs-poolController/files/10714470/replay.2.zip)

**Reply by rstrouse** [reply]
Thanks.  I'll have a look.  While IntelliTouch will not allow you to run both at the same time, njsPC can manipulate a feature that will turn a valve outside of the control of IntelliTouch.  Unfortunately, you are a bit boxed in with IntelliTouch heater control.

I noticed that you do not have Solar selected as the valve circuit for Valve A.  Did you remove this or is your firmware simply assuming Valve A rotation when calling for solar heat?  Check under valves for IntelliTouch configuration.  When I was running IntelliTouch it required that the Solar virtual circuit was selected to turn the valve.  If you can have the Solar heater enabled where IntelliTouch does not move the valve we can make it move whenever there is a call for heat using a rule that turns on a feature attached to that valve.

Please test to see whether the valve moves when nothing is selected for the circuit and there is a call for heat.  Unfortunately, there are differences between versions of the firmware for IntelliTouch for little things like this and the valve configuration message from your panel report that there is nothing connected to Valve A.

Also, pull njsPC since I fixed a cleanup issue with the valve reporting.  There should have been both an Intake and Return valve listed in your valves.

**Reply by antamy** [reply]
I think the system is assuming Valve A is for solar.  Not sure how that is set up through the panel.  I cannot actually change Valve A through the display.  Under Equipment > Solar, Water Solar Present is set to Yes..  This is what I see in the Valves configuration:

![PXL_20230211_175910672](https://user-images.githubusercontent.com/6841038/218277920-39cc38c7-a645-4546-a4dc-1668dab35239.jpg)

I tried a git pull, and restarted. No change to the valves, I just see an Intake

![image](https://user-images.githubusercontent.com/6841038/218277959-0529ff91-caba-42be-b7f4-a6f2b698701a.png)

**Reply by rstrouse** [reply]
You are correct.  IntelliTouch has hardcoded this in the firmware.   The only other way to make this happen is to move your solar valve to valve B.  Then we can add a rule that activates when there is a call for heat and the heat mode is either solar or solar pref.

**Reply by rstrouse** [reply]
The first time it hears a valve configuration message it will add the return valve.  I also added code to pre-initialize the valves on startup like we do for IntelliCenter as this is deterministic at startup.  The only thing we don't yet know about is which circuit is in control of the valve.

**Reply by antamy** [reply]
> You are correct. IntelliTouch has hardcoded this in the firmware. The only other way to make this happen is to move your solar valve to valve B. Then we can add a rule that activates when there is a call for heat and the heat mode is either solar or solar pref.

Ok, this sounds promising.  The only options I have for Solar is for Water Solar or a Heat Pump.  If I set it to Heat Pump, then Valves A and B show as Unused. So there may be an option there.  I can easily move the wires from Valve A to Valve B in the panel and keep the Water Solar configuration.

I'm not clear on your comments re: what circuit is controlling the valve.  Are you referring to a virtual circuit in the software or the actual physical circuit?

Where would the rules run?  I use Home Assistant too, which is obviously quite capable from a rules perspective.  My ideal would be to have the ability to turn on Dual Heating whilst still retaining the existing functionality on the Intellitouch (even if some of that had to move to rules).  In particular, I need to also increase the pump speed when the solar is on.

**Reply by rstrouse** [reply]
I too run home assistant but it is no place for rules that would control an item like this.  I think your best bet is to use Valve B.  There is a rules processor built into the system and I can throw together a .json file for you to drive the rule.  It gives full programmatic control of the events that occur in njsPC and operates regardless of the client.

**Reply by antamy** [reply]
> I too run home assistant but it is no place for rules that would control an item like this. I think your best bet is to use Valve B. There is a rules processor built into the system and I can throw together a .json file for you to drive the rule. It gives full programmatic control of the events that occur in njsPC and operates regardless of the client.

If you could give me an example, that would be great and I can play around with it.

**Reply by rstrouse** [reply]
First create a feature circuit in your IntelliTouch.  Unlike AUX circuits these do not need or control a relay.  In this case all we are doing here is triggering this feature whenever there is a call for any heat on either body. 

Next attach this feature to Valve B.  What this does is rotate the valve whenever that feature is on.  The result is that the valve will rotate whenever that feature is activated.

Next take the json below and save it to a .json file.  This is the action that will be performed when the temps event is triggered.  The temps event is triggered whenever there is a change to the heater functions for the pool or spa.  Change the solarFeatureId in the vars section to the id of the feature created above.

```
{
  "context": {
    "name": "myPoolRules",
    "options": {},
    "vars": {
      "solarFeatureId": 7
    }
  },
  "events": [
    {
      "name": "temps",
      "description": "Process alt solar valve when temps change",
      "processor": [
        "try {",
        "let solarOn = false;",
        "for(let i = 0; i < data.bodies.length;i++) {",
        "let body = data.bodies[i];",
        "if(typeof body.heatStatus !== 'undefined' && (body.heatStatus.name === 'heater' || body.heatStatus.name === 'solar')) solarOn = true;",
        "}",
        "await sys.board.circuits.setFeatureStateAsync(vars.solarFeatureId, solarOn);",
        "} catch(err) { logger.error(`Error triggering solar feature ${err.message}`); }"
      ]
    }
  ]
}

```


Finally, using the file you created above, upload this binding to njsPC.  To do this under the gears menu, select the Controller --> Interfaces Tab.  Then click Add Interface.  Using the Upload Bindings dialog choose the file that you created in the earlier step to upload it to the njsPC server.

Provide the name for your integration select the Type as `Rule` and choose your custom binding from the dropdown.  Then click Add Interface.  This will add a rule interface to the software that you can now enable that will change the valve position whenever there is a call for either heater.

![image](https://user-images.githubusercontent.com/47839015/218326075-c9564fc5-28fa-4ea4-bce3-fb82769b0849.png)

**Reply by antamy** [reply]
Thank you, I'll give this a whirl over the next couple of days.

**Reply by matthewusf** [reply]
I'm trying to eventually do something a little more than this: use the air sensor AS the solar sensor (solar sensor internals are broken) - as well as run the heater and solar simultaneously. However, for now I'm just trying to get this code working as is so that I have a place to start. I have solar in intellicenter turned off, I have a feature (ID 131 per poolConfig.json) which correctly actuates the valve, but with this rule enabled, I get the following error:

[28/03/2023, 11:18:52] info: [11:18:52] 192.168.1.213 PUT /state/body/heatMode {"id":1,"mode":2}
[28/03/2023, 11:19:42] error: Error triggering solar feature Cannot read properties of undefined (reading 'length')
[28/03/2023, 11:19:47] info: [11:19:47] 192.168.1.213 GET /config/body/1/heatModes?null {}
[28/03/2023, 11:19:48] info: [11:19:48] 192.168.1.213 PUT /state/body/heatMode {"id":1,"mode":1}
[28/03/2023, 11:20:00] info: [11:20:00] 192.168.1.213 GET /config/all?null {}
[28/03/2023, 11:20:00] info: [11:20:00] 192.168.1.213 GET /config/options/general?null {}
[28/03/2023, 11:20:02] info: [11:20:02] 192.168.1.213 GET /config/options/circuits?null {}
[28/03/2023, 11:20:04] info: [11:20:04] 192.168.1.213 GET /config/options/features?null {}
[28/03/2023, 11:20:08] info: [11:20:08] 192.168.1.213 PUT /config/feature {"id":131,"name":"Solar Valve","type":0,"showInFeatures":false,"freeze":false,"dontStop":true,"eggTimer":1440}
[28/03/2023, 11:20:12] error: Error triggering solar feature Cannot read properties of undefined (reading 'length')
[28/03/2023, 11:20:16] error: Error triggering solar feature Cannot read properties of undefined (reading 'length')

Appreciate your help!

**Reply by tagyoureit** [reply]
Can you submit a [packet capture](https://github.com/tagyoureit/nodejs-poolController/wiki/How-to-capture-all-packets-for-issue-resolution#-dashpanel)?  We will need to step through and see what's causing the errors.

**Reply by matthewusf** [reply]
Thank you! Looks like line 944 in the console log - wish I could better understand this to not have to pull you in, but I'm really just not sure what I'm looking for!

[replay.zip](https://github.com/tagyoureit/nodejs-poolController/files/11091967/replay.zip)

**Reply by matthewusf** [reply]
I was able to control the valve using a put request to the url /state/circuit/setState with the id and state. I'm actually fine with just building this in python as I'm more familiar with it, but I can't figure out how to "get" solar/air sensor status. I see the event in the developer console evt: 'temps' on dashboard.js:333, but I can't figure out what to do with that - how to use it in python to set a variable. Ideally, I'd also get the set point of the pool.

I'd really like to do a variable speed solar controller that changes pump speed based on the roof sensor - I think I could tackle that in python if I can figure out how to pass these back and forth...

In any case, I'm sure I can figure out json if I can get it to work and better understand what everything is doing...

Appreciate your help!

**Reply by tagyoureit** [reply]
> [28/03/2023, 11:18:52] info: [11:18:52] 192.168.1.213 PUT /state/body/heatMode {"id":1,"mode":2}

This won't work because mode 2 is solar and you have it disabled.  Your workaround for using the circuit will work for turning on/off the valve, but you will not be able to have a solar heat mode and hence no solar temperature.  You really should spend the $30 for a [new one](https://www.amazon.com/Pentair-520272-Temperature-Replacement-Automation/dp/B001DO18FS) and replace the broken one.  

The developer console events you are seeing are Socket.IO events.  The data structures will be the same, but you will want the REST calls.  EG `http://localhost:4200/state/temps`.  Again, this won't have the solar temp since it's disabled.  The API's are somewhat documented in [nodejs-poolcontroller-api](https://tagyoureit.github.io/nodejs-poolcontroller-api/) but the documentation needs to be updated.  You can at least get a most of the calls from here and it should have what you need.

**Reply by matthewusf** [reply]
That's an embarrassing oversight - I did not know that mode 2 was solar. This whole interface is brand new to me so I feel like I'm learning a lot about some parts of this and then still feel so elementary in so many others. I will enable and test that tomorrow - appreciate your help in clarifying where I'm missing the obvious. 

To your point about the sensor: apologies for not being more clear, but it's not my solar sensor that's damaged - it's the actual onboard processing of that sensor that's somewhere messed up. During hurricane Ian, my solar sensor wire was almost cut in half in such a way that it appeared to be shorting the wires, causing an error message in intellicenter. I soldered a new sensor onto new wiring and even tried connecting a 3rd replacement directly into the terminals - but the same error was present. I also factory reset intellicenter to no avail. 

However, the actual sensor disconnected from intellicenter reads the proper resistance to temperature (i.e. 11k at 73F). What finally convinced me is that when connecting that same sensor to either the air or water terminals, those resistance measurements correctly correspond to what is displayed in intellicenter. However, when connected to the solar terminals, as soon as solar is enabled, the displayed temperature quickly spikes to 4XXF to 6XXF to 800F+ and then "ERR". This leads me to believe there is a short or messed up circuit element somewhere - but it's definitely not in the sensor itself. I pulled the board from the head unit and followed the PCB traces but I just don't see anything obvious like a blown capacitor... that said, I don't know much about it so it's more than possible I'm missing something that someone more experienced could fix. 

I now have it connected to the air sensor terminals where it has measured temps ranging from 65 to 130+ on the roof - consistent with what I saw before the damage, so my current plan is to use that sensor in some kind of software controller. That said, I'm more than willing to be corrected anywhere I'm not thinking clearly here - I just want it to work properly. 

Last year, I did have a number of times that I wanted both the solar and heater to run (before an afternoon pool party) - so this post is quite relevant to that (rstrouse sent me to this post from TFP). For now, I'm currently just running Valve A from 11am to sunset but I know that a sensor based approach would be far better and would maybe even allow me to maximize efficiency based on roof temp!

I'll look at the localhost:4200/state/temps tomorrow to see if I can get that to work, but please don't hesitate to critique my approach as I readily admit I'm hacking this together the best I can... Sincerely appreciate you working through this with me!

**Reply by tagyoureit** [reply]
Not embarrassing at all!  The good news is you can retrieve all the relevant values from an equivalent options api endpoint.  For `/state/body/heatmode`, you can call `/config/options/heaters` which will include the correct enumerations for `heatModes`.  

And it sure does sound like you have a resistor or something that's shot along the way.  You might want to try adding additional resistance (maybe 1/10th, 1/20th, ?) and see if you could scale it down, but that's a long shot.   

If you figure out what you want to do, we can help you create a built in integrations rule ([see this example](https://github.com/tagyoureit/nodejs-poolController/discussions/500#discussioncomment-3012759)).  It could be something like:
* If the heat setpoint is > pool temp,
* and air temp is > +3 heat setpoint,
* then turn on solar valve,
* otherwise, turn off solar valve.

You wouldn't need an entire external python app just to read the values and then return an inbound API to make the magic happen.

**Reply by matthewusf** [reply]
Just today finding some time to jump back on this - especially now that the pool is regularly getting above 90 and really needs to have a proper solar controller. Because I'm just more familiar with python, I spent my lunch putting together a script that does exactly what you describe. It works perfectly - thank you so much!

I went ahead and also added a variable speed controller element to it - such that the pump GPM varies based on the differential between the solar temp and the water temp. When it's a higher differential (say 130 and 70), the pump runs at whatever I configure as the maximum (I have it at 50 which is around 55GPM via flowvis). When it's say at the lowest differential to turn on the pump (I have set to 6), then it runs at 25gpm (about 30 via flowvis). 

This turned on/off solar throughout the day with the clouds/rain and correspondingly printed GPM to terminal beautifully, but I haven't figured out how to actually send those GPM commands to intellicenter via any of the the API functions you linked to. Ideally, I would just modify the "Solar Pump Speed" circuit, which I can successfully activate via its ID 134 - but I do not know how to actually change the value of it.

I won't be upset if you tell me this just hasn't been put together because nobody actually **needs** to do this - this part was just a fun extra that I thought might both save a little money while also maximizing heating efficiency during the hottest part of the day. I try to minimize cost by running solar right around 40GPM... but in the hottest parts of a sunny day, the water coming out is much hotter than even our 400k gas heater can get it - so I figure I'll have some benefit in reducing the heat lost by increasing the pump speed.

**Reply by antamy** [reply]
> This turned on/off solar throughout the day with the clouds/rain and correspondingly printed GPM to terminal beautifully, but I haven't figured out how to actually send those GPM commands to intellicenter via any of the the API functions you linked to. Ideally, I would just modify the "Solar Pump Speed" circuit, which I can successfully activate via its ID 134 - but I do not know how to actually change the value of it.
> 
I have Intellitouch and an Intelliflo VS pump.  I can't actually control the pump speeds, except by setting custom pump speeds with names that match the circuits I have set up.  So when Pool is on, my pump runs at the Pool speed, when Spa is on at the Spa speed, Solar at the Solar speed.  Maybe it's the same with your setup?

**Reply by matthewusf** [reply]
It's the same with Intellicenter - but you only get 8 pump circuit assignments and I already use them all. My idea was to just use the one "solar" circuit and dynamically change the GPM/RPM in the same way that you can edit it by going to the "pumps" screen and editing it there - so there must be a function that nodejs is tapping into, I just don't understand how all of this works beyond a very elementary level...

**Reply by antamy** [reply]
Yes, I'm trying to figure it out too!  Maybe an API can be added just to change the value of one pump setting.


That said, there is some useful info here https://www.troublefreepool.com/threads/intellicenter-circuits-pump-and-valve-associations.213335/ about changing speeds and some gotchas.

**Reply by matthewusf** [reply]
Yes @antamy, I was familiar with features/groups/valves and their various associations - but I was hoping for a true "variable" speed option. For now, I just have it running 25, 35, 45, and 55gpm features based on which one matches the GPM output of the program.

This does cause other issues though. For example, I have two skim circuit groups at 45 GPM for 1hr or 12hr (mixing chems or skimming a lot before a get together). When solar gets to the 45 GPM threshold, it activates this same feature circuit, which is then disabled when temps drop the output to 35 GPM. Similarly, if you had skim selected and it hits its egg timer, it will turn off 45 GPM even though solar is calling for it.

I could constantly check the status of the pump to make sure a circuit doesn't turn off and also set up some flags i.e. "alreadyOn45" before turning 45 GPM on, so that when you go to turn it off under 35 GPM, I could check for this flag being 1 or 0 (if 1, don't turn 45 off but then turn the flag off and if 0 then turn 35 off)... but I don't want to hack together something that ideally shouldn't be a permanent solution anyway.

Seems like if nodejsPC can do it, there's gotta be a way for me to also edit a pump circuit speed such that I can just label it "Solar Speed" and then change the GPM via commands like everything else. This works manually quite well, so I really just need a way to edit it.

Any thoughts @tagyoureit or @rstrouse?

![Solar Config](https://user-images.githubusercontent.com/10775620/235498181-14f421e8-6602-4b9b-8c2b-238701e1cac5.jpg)
![Solar (2)](https://user-images.githubusercontent.com/10775620/235498187-42a1856e-7ef7-4dc7-8bf8-59728e01c2b1.jpg)
![Solar (3)](https://user-images.githubusercontent.com/10775620/235498189-bff13ca3-ed39-42c7-954b-fb1de7bf08af.jpg)
![Solar (4)](https://user-images.githubusercontent.com/10775620/235498190-067642a1-20f8-4253-b80e-a4ed1358f487.jpg)
![Solar (5)](https://user-images.githubusercontent.com/10775620/235498191-7e829ab1-4c18-4168-b121-502307cff149.jpg)

**Reply by rstrouse** [reply]
I will look into adding the ability to set a single circuit speed for the pump but why have you not set the solar virtual circuit to your minimum speed to run the solar?  You must know that IntelliCenter, IntelliTouch, and EasyTouch will not accept a single change to only one pump circuit.  It has to have the entire pump configuration sent to it.

The only panel that used to operate like you are requesting was the old Jandy PDA interface.  However, the speed increases on the pump were so slow that mashing the button a couple of times never resulted in your desired speed.  Typically you would set it up so that when certain equipment is on it will run at the highest speed required for the equipment to operate.  For instance, when only the pool is on the speed is 1950 then if there is a call for solar it boosts the speed to 2750.  Once the pool reaches the setpoint it will drop back to 1950 unless the cleaner is also on then it will be 2650 or if the spillway is on then it will be 2280.

![image](https://user-images.githubusercontent.com/47839015/235535499-38ac9ac8-b60c-4082-ba45-d44a9b3dc757.png)

**Reply by matthewusf** [reply]
_> I will look into adding the ability to set a single circuit speed for the pump but why have you not set the solar virtual circuit to your minimum speed to run the solar?_ 
Well this is definitely not necessary, but I was reading about how heater efficiency is directly tied to heat loss due to the temperature differential between the heated water and the surrounding air. My understanding is that panels are rated for a certain GPM range because, at the expected hottest temperatures with the maximum temperature differential between the panel and outside air, you're minimizing heat loss if you stay within that flow range. So if it's 80 outside and the water temperature is 80 going in and 120 coming out at 35gpm but increasing it to 55gpm brings that temperature down to 90, then more heat is successfully transferred because less is lost. Obviously the other side of the coin is that it costs more to run the pump at a higher speed, but the flip side is also true - my "stepped" solar controller is running it at 25gpm until the afternoon heat while the temperature differential is still around 10 degrees, saving me money most of the day but increasing the speed when more heat is available.

So my goal is just to estimate the pump speed at which the temperature differential stays the same, up to a certain point where the pump is too expensive to run relative to the heat gained. Just a fun side project that might save a few bucks and increase heat returned to the pool a bit.


_> You must know that IntelliCenter, IntelliTouch, and EasyTouch will not accept a single change to only one pump circuit. It has to have the entire pump configuration sent to it._
I did not know that - so whenever I make a change to a pump speed in nodejspoolControllerdashPanel, it is required to send a whole new pump configuration of all circuits and their respective speeds through? I can imagine that's quite a hassle and understand if this isn't something you want to make as a priority! Just let me know if you have any other thoughts on it...

**Comment by rstrouse** [comment]
Oh and one other thing.  Do not run where you are logging all the packets to a file indefinitely.  This will eventually build a huge file.  Uncheck that log to file for normal operation.  When you do capture replay you do not need to turn on the logging as njsPC will handle all of that for you.  The logging filters are simply used to zero in on specific equipment or intermittent issues.

**Comment by antamy** [comment]
I'm keeping an eye on this to see what you come up with. I still want to get this working at my house.  I'll be able to collaborate on it in about 3 weeks time.



________________________________

**Comment by antamy** [comment]
@matthewusf - have you had success with this?  I am thinking of a different approach to enabling a "boost mode" for heating, which would use a small circuit and Valve B.    

When I have solar activated, that also speeds up my pump, so I don't want to have to replicate logic for controlling solar and the pump.  I also don't want to lose any functionality on the Intellitouch display.   My thought is to use Valve B as the switch for Boost mode.  In the software / automation, I would need a way to enable Boost mode.  Once enabled, if Boost is ON and Solar is ON, then Valve B would be made active.  This would then activate a relay or MOSFET to turn on the heater.  If the Heater pin on the control board is on, that would activate a relay or MOSFET or transistor to supply power to the heater.  This combo relay / MOSFET circuit would sit between the gas heater and Valve B pins on the board and the wire controlling the heater.  In my setup, my heat pump heater is controlled through the gas heater connection on the board.

What I don't know (since I'm1,300 miles away from the panel) are the voltages.  I think Valve B is 24v and would assume the heater pin is 24V too.  If you know what these voltages or have any thoughts on feasibility of my approach, let me know!

**Reply by tagyoureit** [reply]
The heater connection is a fireman's switch.  Do NOT under any circumstance apply voltage to that connection point.  @rstrouse did an excellent job in the [Nixie wiki](https://github.com/tagyoureit/nodejs-poolController/wiki/DIY-Standalone-Nixie-Pool-Controller) documenting all of the voltages.  Valves are 24vac.

**Reply by antamy** [reply]
> The heater connection is a fireman's switch. Do NOT under any circumstance apply voltage to that connection point. @rstrouse did an excellent job in the [Nixie wiki](https://github.com/tagyoureit/nodejs-poolController/wiki/DIY-Standalone-Nixie-Pool-Controller) documenting all of the voltages. Valves are 24vac.

This makes the hardware very simple - just a relay connected to valve B with the switch side connected into the wire to the heater.  I'll install that in a couple of weeks and give it a whirl!
